lib/logging/include/logging/logging.h

```diff
@@ -0,0 +1,81 @@
+ */
+
+#ifndef HSE_LOGGING_LOGGING_H
+#define HSE_LOGGING_LOGGING_H
+
+#include <hse_util/compiler.h>
+#include <hse_util/event_counter.h>
+
+#include <error/merr.h>
+
+#include "types.h"
+
+/* clang-format off */
+
+#ifndef LP_DEFAULT
+#define LP_DEFAULT LP_DEBUG
+#endif
+
+#ifdef HSE_RELEASE_BUILD
+#define HSE_LOG_SQUELCH_NS_DEFAULT (1000 * 1000)
+#else
+#define HSE_LOG_SQUELCH_NS_DEFAULT (0)
+#endif
+
+#define log_pri(_pri, _err, _fmt, ...)                         \
+    do {                                                       \
+        static struct event_counter hse_ev_log _dt_section = { \
+            .ev_odometer = 0,                                  \
+            .ev_pri = (_pri),                                  \
+            .ev_flags = EV_FLAGS_HSE_LOG,                      \
+            .ev_file = __FILE__,                               \
+            .ev_line = __LINE__,                               \
+            .ev_dte = {                                        \
+                .dte_data = &hse_ev_log,                       \
+                .dte_ops = &event_counter_ops,                 \
+                .dte_type = DT_TYPE_ERROR_COUNTER,             \
+                .dte_line = __LINE__,                          \
+                .dte_file = __FILE__,                          \
+                .dte_func = __func__,                          \
+            }                                                  \
+        };                                                     \
+                                                               \
+        log_impl(&hse_ev_log, (_err), (_fmt), ##__VA_ARGS__);  \
+    } while (0)
+
+/* Main logging APIs for HSE.  Debug, info and warn levels use async
+ * logging. Err and crit levels use synchronous logging.
+ */
+#define log_debug(_fmt, ...)    log_pri(LP_DEBUG, 0, (_fmt), ##__VA_ARGS__)
+#define log_info(_fmt, ...)     log_pri(LP_INFO, 0, (_fmt), ##__VA_ARGS__)
+#define log_warn(_fmt, ...)     log_pri(LP_WARN, 0, (_fmt), ##__VA_ARGS__)
+#define log_err(_fmt, ...)      log_pri(LP_ERR, 0, (_fmt), ##__VA_ARGS__)
+#define log_crit(_fmt, ...)     log_pri(LP_CRIT, 0, (_fmt), ##__VA_ARGS__)
+
+/* Emit logs with pretty printed merr_t values
+ */
+#define log_warnx(_fmt, _err, ...)  log_pri(LP_WARN, (_err), (_fmt), ##__VA_ARGS__)
+#define log_errx(_fmt, _err, ...)   log_pri(LP_ERR, (_err), (_fmt), ##__VA_ARGS__)
+
+merr_t
+logging_init(void) HSE_COLD;
+
+void
+logging_fini(void) HSE_COLD;
+
+/* Helper API used by above log macros. Not intended to be called directly. */
+void
+log_impl(struct event_counter *ev, merr_t err, const char *fmt, ...) HSE_PRINTF(3, 4);
+
+/* Convert a log priority level: numeric to string. */
+const char *
+log_priority_to_string(enum log_priority val);
+
+/* Convert a log priority level: string to numeric. */
+enum log_priority
+log_priority_from_string(const char *name);
+
+#endif /* HSE_LOGGING_LOGGING_H */
```

Comment:
What did moving the file around buy us?

Reply:
I generally think internal static libraries lead to better relationships between dependent pieces of code. `util` has a lot of code that I think belongs in its own static library like event counters, perf counters, and workqueues for a few examples. I also think these smaller static libraries lead to more maintainable code.